go/libkb/device_clone_state.go

```diff
@@ -0,0 +1,191 @@
+// Copyright 2018 Keybase, Inc. All rights reserved. Use of
+// this source code is governed by the included BSD license.
+
+package libkb
+
+import (
+	"fmt"
+
+	jsonw "github.com/keybase/go-jsonw"
+)
+
+const (
+	DefaultCloneTokenValue string = "00000000000000000000000000000000"
+)
+
+type DeviceCloneState struct {
+	Prior  string
+	Stage  string
+	Clones int
+}
+
+type DeviceCloneStateJSONFile struct {
+	*JSONFile
+}
+type cloneDetectionResponse struct {
+	Status AppStatus `json:"status"`
+	Token  string    `json:"token"`
+	Clones int       `json:"clones"`
+}
+
+func (d *cloneDetectionResponse) GetAppStatus() *AppStatus {
+	return &d.Status
+}
+
+func UpdateDeviceCloneState(m MetaContext) (before int, after int, err error) {
+	d, err := GetDeviceCloneState(m)
+	before = d.Clones
+
+	prior, stage := d.Prior, d.Stage
+	if prior == "" {
+		//first run
+		prior = DefaultCloneTokenValue
+	}
+	if stage == "" {
+		stage, err = RandHexString("", 16)
+		if err != nil {
+			return 0, 0, err
+		}
+		err = SetDeviceCloneState(m, DeviceCloneState{Prior: prior, Stage: stage, Clones: d.Clones})
+		if err != nil {
+			return 0, 0, err
+		}
+	}
+
+	// POST these tokens to the server
+	arg := APIArg{
+		Endpoint:    "device/clone_detection_token",
+		SessionType: APISessionTypeREQUIRED,
+		Args: HTTPArgs{
+			"device_id": m.G().ActiveDevice.DeviceID(),
+			"prior":     S{Val: prior},
+			"stage":     S{Val: stage},
+		},
+		MetaContext:    m,
+		AppStatusCodes: []int{SCOk},
+	}
+	var res cloneDetectionResponse
+	err = m.G().API.PostDecode(arg, &res)
+	if err != nil {
+		return 0, 0, err
+	}
+
+	err = SetDeviceCloneState(m, DeviceCloneState{Prior: res.Token, Stage: "", Clones: res.Clones})
+	after = res.Clones
+	return
+}
+
+func configPaths(m MetaContext) (p, s, c string) {
+	deviceID := m.G().ActiveDevice.DeviceID()
+	p = fmt.Sprintf("%s.prior", deviceID)
+	s = fmt.Sprintf("%s.stage", deviceID)
+	c = fmt.Sprintf("%s.clones", deviceID)
+	return
```

Comment:
`return p,s,c`;  we try to never do "naked" returns anymore.

Reply:
cool

Reply:
Yeah, I think the danger of "naked returns" comes with refactoring something higher up in the function, and then forgetting there's a naked return hanging out below.

Reply:
yeah - that makes perfect sense.